larq_compute_engine/tflite/tests/bconv2d_test.cc

```diff
@@ -386,6 +496,88 @@ class BConv2DOpTest : public ::testing::TestWithParam<TestParamTuple> {
   }
 };
 
+// Useful struct for int8 quantization
+template <typename T>
+struct LceTensor : public TensorData {
+  LceTensor(std::vector<int> shape = {})
+      : TensorData(GetTensorType<T>(), shape), reciprocal_scale(1){};
+
+  template <typename TOther>
+  void SetQuantizationParams(const LceTensor<TOther>& rhs) {
+    reciprocal_scale = rhs.reciprocal_scale;
+    scale = rhs.scale;
+    zero_point = rhs.zero_point;
+  }
+
+  // For int8, `int8_value` has range [-128, 127]
+  //     real_value = scale * (int8_value - zero_point)
+  // To represent {-1,+1} without error, we require
+  //     int_value = zero_point + (+-1) / scale
+  // Therefore `scale` must be of the form 1/n
+  // and we require
+  //     abs(zero_point) + abs(1/scale) <= 127
+  // Note that filter.zero_point = 0 for Conv2D,
+  // and BConv2D does not have int8 filters.
+
+  template <typename Gen>
+  void GenerateQuantizationParams(Gen& gen) {
+    reciprocal_scale = std::uniform_int_distribution<>(1, 20)(gen);
+    scale = 1.0f / float(reciprocal_scale);
+    zero_point = std::uniform_int_distribution<>(-20, 20)(gen);
+  }
+
+  // This has zero_point = 0
+  template <typename Gen>
+  void GenerateQuantizationParamsPerChannel(Gen& gen) {
+    reciprocal_scale = std::uniform_int_distribution<>(1, 20)(gen);
+    // Do not set this->scale because TF Lite might expect it to be 0 when
+    // per_channel_quantization is used.
+    float s = 1.0f / float(reciprocal_scale);
+
+    per_channel_quantization = true;
+    std::size_t count = shape[channel_index];
+    per_channel_quantization_scales = std::vector<float>(count, s);
+    per_channel_quantization_offsets = std::vector<std::int64_t>(count, 0);
+  }
+
+  T Quantize(const std::int32_t x) const {
```

Comment:
For our tests we only allow scales of the form `1/n` so that we can represent `-1` and `+1` without error. Therefore we have some specialized `Quantize` functions that are not available in TF code.